### PR TITLE
Add System.Object converter sample to have semantics similar to Newtonsoft Json.NET

### PR DESCRIPTION
**Pull request by steveharter**
Sample converter and tests for converting JSON to System.Object-declared properties using similar semantics as Json.NET.
- TokenType.True -> System.Bool (true)
- TokenType.False-> System.Bool (false)
- TokenType.Number -> System.Long or System.Double
- TokenType.String -> System.String or System.DateTime
- TokenType.Null -> null
- Other (object, array) -> JsonElement (not the same)

related to discussion from https://github.com/dotnet/corefx/issues/38713

### CODE REVIEW

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.Object.cs [hunk 1]
```diff
@@ -259,5 +259,198 @@ private class TestClassWithFieldsHavingCustomConverter
             public int IntValue { get; set; }
             public string Message { get; set; }
         }
+
+        /// <summary>
+        /// A converter that converts a JSON "true" and "false" tokens to a System.Bool,
+        /// a JSON number to a System.Double, and a JSON string to either System.String (if not a date)
+        /// or System.DateTimeOffset or System.Date depending on whether the date has an offset.
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/CustomConverterTests.Object.cs]:
```suggestion
        /// or System.DateTimeOffset or System.DateTime depending on whether the date has an offset.
```

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.Object.cs [hunk 2]
```diff
@@ -259,5 +259,198 @@ private class TestClassWithFieldsHavingCustomConverter
...
+                }
+
+                if (reader.TokenType == JsonTokenType.Number)
+                {
+                    return reader.GetDouble();
+                }
+
+                if (reader.TokenType == JsonTokenType.String)
+                {
+                    if (reader.TryGetDateTime(out DateTime datetime))
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/CustomConverterTests.Object.cs]: Should we always return a DateTimeOffset instead?

@GrabYourPitchforks - you had thoughts about this previously. What do you think?

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.Object.cs [hunk 3]
```diff
@@ -259,5 +259,198 @@ private class TestClassWithFieldsHavingCustomConverter
...
+                {
+                    return false;
+                }
+
+                if (reader.TokenType == JsonTokenType.Number)
+                {
+                    return reader.GetDouble();
+                }
+
+                if (reader.TokenType == JsonTokenType.String)
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/CustomConverterTests.Object.cs]: What about Guid?

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.Object.cs [hunk 4]
```diff
@@ -259,5 +259,198 @@ private class TestClassWithFieldsHavingCustomConverter
...
+
+                {
+                    string json = JsonSerializer.Serialize<object>(new int[] { }, options);
+                    Assert.Equal("[]", json);
+                }
+            }
+
+            // Results should be the same with or without the custom converter since the
+            // serializer calls value.GetType() for every property value declared as System.Object.
+            Verify(new JsonSerializerOptions());
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/CustomConverterTests.Object.cs]: Nice!

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.Object.cs (hunk 2 above)
Comment (steveharter) [reply src/System.Text.Json/tests/Serialization/CustomConverterTests.Object.cs]: I was basing it off the issue, which suggested DateTime when no offset\timezone is specified.

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.Object.cs (hunk 3 above)
Comment (steveharter) [reply src/System.Text.Json/tests/Serialization/CustomConverterTests.Object.cs]: I could add it... Maybe cross-reference with Json.NET?

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.Object.cs (hunk 2 above)
Comment (steveharter) [reply src/System.Text.Json/tests/Serialization/CustomConverterTests.Object.cs]: Note I changed the PR to have Json.NET semantics which uses DateTime only.

File: src/System.Text.Json/tests/Serialization/CustomConverterTests.Object.cs (hunk 3 above)
Comment (steveharter) [reply src/System.Text.Json/tests/Serialization/CustomConverterTests.Object.cs]: Now basing semantics on Json.NET which doesn't do anything with Guid.

**Comment by steveharter** [comment]
Test failure in Windows x86_release not related (System.Diagnostics.EventLog.Tests)

**Pull request merged**